File: task3-unknown-language/Kotlin-notes-app/src/main/kotlin/com/notes/service/NoteService.kt
package com.notes.service

import com.notes.model.Note
import com.notes.repository.NoteRepository

/**
 * Исключение для ошибок валидации
 */
class ValidationException(message: String) : Exception(message)

/**
 * Исключение для случаев, когда заметка не найдена
 */
class NoteNotFoundException(message: String) : Exception(message)

/**
 * Сервис для работы с заметками, содержит бизнес-логику и валидацию
 */
class NoteService(private val noteRepository: NoteRepository) {
    
    companion object {
        private const val MIN_TEXT_LENGTH = 1
        private const val MAX_TEXT_LENGTH = 500
    }
    
    /**
     * Получить все заметки
     * @return список всех заметок
     */
    fun getAllNotes(): List<Note> {
        return noteRepository.findAll()
    }
    
    /**
     * Получить заметку по ID
     * @param id идентификатор заметки
     * @return заметка
     * @throws NoteNotFoundException если заметка не найдена
     */
    fun getNoteById(id: Long): Note {
        return noteRepository.findById(id) 
            ?: throw NoteNotFoundException("Заметка с ID $id не найдена")
    }
    
    /**
     * Создать новую заметку
     * @param text текст заметки
     * @return созданная заметка
     * @throws ValidationException если текст не прошел валидацию
     */
    fun createNote(text: String): Note {
        validateNoteText(text)
        return noteRepository.create(text.trim())
    }
    
    /**
     * Обновить заметку
     * @param id идентификатор заметки
     * @param text новый текст заметки
     * @return обновленная заметка
     * @throws ValidationException если текст не прошел валидацию
     * @throws NoteNotFoundException если заметка не найдена
     */
    fun updateNote(id: Long, text: String): Note {
        validateNoteText(text)
        
        return noteRepository.update(id, text.trim()) 
            ?: throw NoteNotFoundException("Заметка с ID $id не найдена")
    }
    
    /**
     * Удалить заметку
     * @param id идентификатор заметки
     * @throws NoteNotFoundException если заметка не найдена
     */
    fun deleteNote(id: Long) {
        val deleted = noteRepository.delete(id)
        if (!deleted) {
            throw NoteNotFoundException("Заметка с ID $id не найдена")
        }
    }
    
    /**
     * Валидация текста заметки
     * @param text текст для валидации
     * @throws ValidationException если текст не соответствует требованиям
     */
    private fun validateNoteText(text: String) {
        val trimmedText = text.trim()
        
        when {
            trimmedText.isEmpty() -> {
                throw ValidationException("Текст заметки не может быть пустым")
            }
            trimmedText.length < MIN_TEXT_LENGTH -> {
                throw ValidationException("Текст заметки должен содержать минимум $MIN_TEXT_LENGTH символ")
            }
            trimmedText.length > MAX_TEXT_LENGTH -> {
                throw ValidationException("Текст заметки не может превышать $MAX_TEXT_LENGTH символов")
            }
        }
    }
    
    /**
     * Проверить существование заметки
     * @param id идентификатор заметки
     * @return true если заметка существует, false в противном случае
     */
    fun noteExists(id: Long): Boolean {
        return try {
            noteRepository.findById(id) != null
        } catch (e: Exception) {
            false
        }
    }
}

File: task3-unknown-language/Rust_notes_app/src/services/note_service.rs
use crate::models::{Note, CreateNoteRequest, UpdateNoteRequest, validate_note_text, ValidationError};
use crate::repository::NoteRepository;
use anyhow::Result;
use uuid::Uuid;

#[derive(thiserror::Error, Debug)]
pub enum ServiceError {
    #[error("Заметка не найдена")]
    NotFound,
    #[error("Ошибка валидации: {0}")]
    Validation(#[from] ValidationError),
    #[error("Ошибка базы данных: {0}")]
    Database(#[from] anyhow::Error),
}

#[derive(Clone)]
pub struct NoteService {
    repository: NoteRepository,
}

impl NoteService {
    pub fn new(repository: NoteRepository) -> Self {
        Self { repository }
    }

    pub async fn create_note(&self, request: CreateNoteRequest) -> Result<Note, ServiceError> {
        // Валидация входных данных
        validate_note_text(&request.text)?;

        let note = Note::new(request.text);
        self.repository.create(&note).await?;

        Ok(note)
    }

    pub async fn get_all_notes(&self) -> Result<Vec<Note>, ServiceError> {
        let notes = self.repository.find_all().await?;
        Ok(notes)
    }

    pub async fn get_note_by_id(&self, id: Uuid) -> Result<Note, ServiceError> {
        match self.repository.find_by_id(id).await? {
            Some(note) => Ok(note),
            None => Err(ServiceError::NotFound),
        }
    }

    pub async fn update_note(&self, id: Uuid, request: UpdateNoteRequest) -> Result<Note, ServiceError> {
        // Валидация входных данных
        validate_note_text(&request.text)?;

        // Проверяем, существует ли заметка
        let mut note = match self.repository.find_by_id(id).await? {
            Some(note) => note,
            None => return Err(ServiceError::NotFound),
        };

        // Обновляем заметку
        note.update_text(request.text);
        
        // Сохраняем изменения
        self.repository.update(&note).await?;

        Ok(note)
    }

    pub async fn delete_note(&self, id: Uuid) -> Result<(), ServiceError> {
        let deleted = self.repository.delete(id).await?;
        
        if !deleted {
            return Err(ServiceError::NotFound);
        }

        Ok(())
    }
}